lib/plugins/aws/deploy/compile/functions/index.js

```diff
@@ -17,15 +17,44 @@ class AwsCompileFunctions {
     };
   }
 
-  compileRole(role) {
-    if (typeof role === 'object') {
-      // role is an "Fn::GetAtt" object
-      return role;
-    } else if (role.indexOf(':') === -1) {
-      // role is a Logical Role Name
-      return { 'Fn::GetAtt': [role, 'Arn'] };
+  compileRole(newFunction, role) {
+    const compiledFunction = newFunction;
+    const unnsupportedRoleError = new this.serverless.classes
+      .Error(`Unsupported role provided: "${role}"`);
+
+    switch (typeof role) {
+      case 'object':
+        if ('Fn::GetAtt' in role) {
+          // role is an "Fn::GetAtt" object
+          compiledFunction.Properties.Role = role;
+          compiledFunction.DependsOn = [role['Fn::GetAtt'][0]];
+        } else if ('Ref' in role || 'Fn::ImportValue' in role) {
```

Comment:
Ref should be able to reference resources based on the logicalId which are created in the custom resource section.

```yml
functions:
  hello:
    handler: handler.hello
    role:
      Ref: role1

resources:
  Outputs:
    myCustomOutput:
      Value:
        Ref: role1
  Resources:
    role1:
      Type: AWS::IAM::Role
      ...
```